spec/lib/ddf_spec.rb

```diff
@@ -0,0 +1,94 @@
+      :label      => _("Username"),
+      :helperText => _("Should have privileged access, such as root or administrator."),
+      :isRequired => true,
+      :validate   => [{:type => "required"}]
+    }
+  end
+
+  let(:schema) do
+    {
+      :fields => [
+        {
+          :component => 'sub-form',
+          :id        => 'endpoints-subform',
+          :name      => 'endpoints-subform',
+          :title     => _("Endpoint"),
+          :fields    => [
+            {
+              :component              => 'validate-provider-credentials',
+              :id                     => 'authentications.default.valid',
+              :name                   => 'authentications.default.valid',
+              :skipSubmit             => true,
+              :validationDependencies => %w[type zone_id],
+              :fields                 => [
+                {
+                  :component  => "text-field",
+                  :id         => "endpoints.default.url",
+                  :name       => "endpoints.default.url",
+                  :label      => _("URL"),
+                  :isRequired => true,
+                  :validate   => [{:type => "required"}]
+                },
+                {
+                  :component    => "select",
+                  :id           => "endpoints.default.verify_ssl",
+                  :name         => "endpoints.default.verify_ssl",
+                  :label        => _("SSL verification"),
+                  :isRequired   => true,
+                  :initialValue => OpenSSL::SSL::VERIFY_PEER,
+                  :options      => [
+                    {
+                      :label => _('Do not verify'),
+                      :value => OpenSSL::SSL::VERIFY_NONE,
+                    },
+                    {
+                      :label => _('Verify'),
+                      :value => OpenSSL::SSL::VERIFY_PEER,
+                    },
+                  ]
+                },
+                userid_field,
+                {
+                  :component  => "password-field",
+                  :id         => "authentications.default.password",
+                  :name       => "authentications.default.password",
+                  :label      => _("Password"),
+                  :type       => "password",
+                  :isRequired => true,
+                  :validate   => [{:type => "required"}]
+                },
+              ],
+            },
+          ],
+        },
+      ]
+    }.freeze
+  end
+
+  it ".extract_attributes" do
+    expect(described_class.extract_attributes(schema, :id)).to eq %w[
+      endpoints-subform
+      authentications.default.valid
+      endpoints.default.url
+      endpoints.default.verify_ssl
+      authentications.default.userid
+      authentications.default.password
+    ]
+  end
+
+  describe ".find_field" do
+    it "when field is found" do
+      expect(described_class.find_field(schema, userid_field[:id])).to equal userid_field
```

Comment:
I used `equal` here on purpose as I want it to show it's finding the actual field itself and not a copy.